Prevent Ingests with Illegal Data

lkiesow opened the pull request:
This patch prevents the ingest service from accepting catalogs
containing data with characters illegal  in XML which would cause
Opencast to fail internally.

Instead, such ingests are now prevented by returning a 400 BAD REQUEST
from the start.

An example for a problematic ingest would be:

    curl -f -i -u admin:opencast \
      http://localhost:8080/ingest/addMediaPackage/fast \
      -F 'flavor=presenter/source' \
      -F mediaUri=https://www.pexels.com/video/854982/download/ \
      -F title=Test \
      -F creator="$(printf "_\xB_\n")"

This is related to #2187 and #2534.

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] be against the correct branch (features can only go into develop)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

lkiesow committed:
Prevent Ingests with Illegal Data

This patch prevents the ingest service from accepting catalogs
containing data with characters illegal  in XML which would cause
Opencast to fail internally.

Instead, such ingests are now prevented by returning a 400 BAD REQUEST
from the start.

An example for a problematic ingest would be:

    curl -f -i -u admin:opencast \
      http://localhost:8080/ingest/addMediaPackage/fast \
      -F 'flavor=presenter/source' \
      -F mediaUri=https://www.pexels.com/video/854982/download/ \
      -F title=Test \
      -F creator="$(printf "_\xB_\n")"

LukasKalbertodt commented on modules/ingest-service-impl/src/main/java/org/opencastproject/ingest/endpoint/IngestRestService.java:
```diff
@@ -1306,6 +1309,22 @@ public Response addDCCatalog(@FormParam("mediaPackage") String mp, @FormParam("d
+   * @param message
+   *          Message to send
+   * @param e
+   *          Exception to log. If <pre>null</pre>, a new exception is created to log a stack tracke.
```
```suggestion
   *          Exception to log. If <pre>null</pre>, a new exception is created to log a stack trace.
```

Arnei commented:
Seems to return `500 Internal Server Error` for me if I try to ingest a custom catalog on the `/ingest/addCatalog` Endpoint.

<details>
  <summary>Catalog</summary>  
  
```
[
  {
    "begin": 0,
    "duration": 73695
  }
]
```

</details>

<details>
  <summary>Error Log</summary>  
  
```
2021-04-27 16:34:32,464 | WARN  | (IngestRestService:562) - Unable to add mediapackage element
java.io.IOException: Catalog XML is invalid
	at org.opencastproject.ingest.impl.IngestServiceImpl.addCatalog(IngestServiceImpl.java:935) ~[?:?]
	at org.opencastproject.ingest.endpoint.IngestRestService.addMediaPackageElement(IngestRestService.java:548) ~[?:?]
	at org.opencastproject.ingest.endpoint.IngestRestService.addMediaPackageCatalog(IngestRestService.java:426) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179) ~[?:?]
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) ~[?:?]
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201) ~[?:?]
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104) ~[?:?]
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59) ~[?:?]
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96) ~[?:?]
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308) ~[?:?]
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121) ~[?:?]
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267) ~[?:?]
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234) ~[?:?]
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208) ~[?:?]
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160) ~[?:?]
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:216) ~[?:?]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:301) ~[?:?]
	at org.opencastproject.kernel.rest.RestPublisher$RestServlet.handleRequest(RestPublisher.java:558) ~[?:?]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:220) ~[?:?]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707) ~[!/:3.1.0]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:276) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1631) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.rest.JsonpFilter.doFilter(JsonpFilter.java:113) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.filter.proxy.TransparentProxyFilter.doFilter(TransparentProxyFilter.java:63) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.security.urlsigning.filter.UrlSigningFilter.doFilter(UrlSigningFilter.java:86) ~[!/:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.security.RemoteUserAndOrganizationFilter.doFilter(RemoteUserAndOrganizationFilter.java:247) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.rest.CurrentJobFilter.doFilter(CurrentJobFilter.java:87) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) ~[?:?]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118) ~[?:?]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.opencastproject.kernel.security.AsyncTimeoutRedirectFilter.doFilter(AsyncTimeoutRedirectFilter.java:60) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:103) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:113) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:146) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:54) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:45) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:98) ~[?:?]
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:201) ~[?:?]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:102) ~[?:?]
	at org.springframework.security.web.authentication.www.DigestAuthenticationFilter.doFilter(DigestAuthenticationFilter.java:115) ~[?:?]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:102) ~[?:?]
	at org.springframework.web.filter.CompositeFilter.doFilter(CompositeFilter.java:82) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.ui.DefaultLoginPageGeneratingFilter.doFilter(DefaultLoginPageGeneratingFilter.java:91) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:183) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:105) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192) ~[?:?]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160) ~[?:?]
	at org.opencastproject.kernel.security.SecurityFilter.doFilter(SecurityFilter.java:124) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.security.OrganizationFilter.doFilter(OrganizationFilter.java:135) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.filter.https.HttpsFilter.doFilter(HttpsFilter.java:73) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.rest.CleanSessionsFilter.doFilter(CleanSessionsFilter.java:89) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:226) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:549) ~[!/:9.4.28.v20200408]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) ~[!/:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1363) ~[!/:9.4.28.v20200408]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:298) ~[!/:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:489) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1278) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[!/:9.4.28.v20200408]
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:82) ~[!/:?]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.Server.handle(Server.java:500) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938) [!/:9.4.28.v20200408]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_242]
Caused by: org.xml.sax.SAXParseException: Content is not allowed in prolog.
	at org.apache.xerces.parsers.DOMParser.parse(Unknown Source) ~[?:?]
	at org.apache.xerces.jaxp.DocumentBuilderImpl.parse(Unknown Source) ~[?:?]
	at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:121) ~[?:1.8.0_242]
	at org.opencastproject.ingest.impl.IngestServiceImpl.addCatalog(IngestServiceImpl.java:932) ~[?:?]
	... 115 more
```
</details>

Arnei commented:
Might not fix dublincore catalogs for series. Silently fails for POST requests against `/series/` with a dublincore containing a vertical tab.

<details>
  <summary>Error Log</summary>  
  
```
2021-04-27 17:03:03,522 | WARN  | (DublinCoreXmlFormat:299) - Fatal error parsing DublinCore catalog: An invalid XML character (Unicode: 0xb) was found in the element content of the document.
2021-04-27 17:03:03,523 | WARN  | (LogUtils:475) - Application {http://endpoint.googlespeech.transcription.opencastproject.org/}GoogleSpeechTranscriptionRestService has thrown exception, unwinding now
org.apache.cxf.interceptor.Fault: Unable to read DublinCore catalog, XML parsing failed.
	at org.apache.cxf.service.invoker.AbstractInvoker.createFault(AbstractInvoker.java:162) ~[?:?]
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:128) ~[?:?]
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201) ~[?:?]
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104) ~[?:?]
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59) ~[?:?]
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96) ~[?:?]
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308) ~[?:?]
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121) ~[?:?]
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267) ~[?:?]
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234) ~[?:?]
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208) ~[?:?]
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160) ~[?:?]
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:216) ~[?:?]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:301) ~[?:?]
	at org.opencastproject.kernel.rest.RestPublisher$RestServlet.handleRequest(RestPublisher.java:558) ~[?:?]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:220) ~[?:?]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707) ~[!/:3.1.0]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:276) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1631) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.rest.JsonpFilter.doFilter(JsonpFilter.java:113) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.filter.proxy.TransparentProxyFilter.doFilter(TransparentProxyFilter.java:63) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.security.urlsigning.filter.UrlSigningFilter.doFilter(UrlSigningFilter.java:86) ~[!/:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.security.RemoteUserAndOrganizationFilter.doFilter(RemoteUserAndOrganizationFilter.java:247) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.rest.CurrentJobFilter.doFilter(CurrentJobFilter.java:87) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) ~[?:?]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118) ~[?:?]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.opencastproject.kernel.security.AsyncTimeoutRedirectFilter.doFilter(AsyncTimeoutRedirectFilter.java:60) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:103) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:113) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:146) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:54) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:45) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:98) ~[?:?]
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:201) ~[?:?]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:102) ~[?:?]
	at org.springframework.security.web.authentication.www.DigestAuthenticationFilter.doFilter(DigestAuthenticationFilter.java:115) ~[?:?]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:102) ~[?:?]
	at org.springframework.web.filter.CompositeFilter.doFilter(CompositeFilter.java:82) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.ui.DefaultLoginPageGeneratingFilter.doFilter(DefaultLoginPageGeneratingFilter.java:91) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:183) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:105) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87) ~[?:?]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) ~[?:?]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192) ~[?:?]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160) ~[?:?]
	at org.opencastproject.kernel.security.SecurityFilter.doFilter(SecurityFilter.java:124) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.security.OrganizationFilter.doFilter(OrganizationFilter.java:135) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.filter.https.HttpsFilter.doFilter(HttpsFilter.java:73) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.opencastproject.kernel.rest.CleanSessionsFilter.doFilter(CleanSessionsFilter.java:89) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:226) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:549) ~[!/:9.4.28.v20200408]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) ~[!/:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1363) ~[!/:9.4.28.v20200408]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:298) ~[!/:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:489) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1278) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[!/:9.4.28.v20200408]
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:82) ~[!/:?]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.Server.handle(Server.java:500) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383) ~[!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806) [!/:9.4.28.v20200408]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938) [!/:9.4.28.v20200408]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_242]
Caused by: java.lang.RuntimeException: Unable to read DublinCore catalog, XML parsing failed.
	at org.opencastproject.metadata.dublincore.DublinCores.read(DublinCores.java:188) ~[?:?]
	at org.opencastproject.metadata.dublincore.DublinCoreCatalogService.load(DublinCoreCatalogService.java:182) ~[?:?]
	at org.opencastproject.series.endpoint.SeriesRestService.addOrUpdateSeries(SeriesRestService.java:423) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179) ~[?:?]
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) ~[?:?]
	... 107 more
Caused by: org.xml.sax.SAXParseException: An invalid XML character (Unicode: 0xb) was found in the element content of the document.
	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source) ~[?:?]
	at org.apache.xerces.util.ErrorHandlerWrapper.fatalError(Unknown Source) ~[?:?]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source) ~[?:?]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source) ~[?:?]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source) ~[?:?]
	at org.apache.xerces.impl.XMLScanner.reportFatalError(Unknown Source) ~[?:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source) ~[?:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source) ~[?:?]
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source) ~[?:?]
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source) ~[?:?]
	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source) ~[?:?]
	at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source) ~[?:?]
	at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source) ~[?:?]
	at org.apache.xerces.jaxp.SAXParserImpl.parse(Unknown Source) ~[?:?]
	at org.opencastproject.metadata.dublincore.DublinCoreXmlFormat.readImpl(DublinCoreXmlFormat.java:247) ~[?:?]
	at org.opencastproject.metadata.dublincore.DublinCoreXmlFormat.read(DublinCoreXmlFormat.java:145) ~[?:?]
	at org.opencastproject.metadata.dublincore.DublinCores.read(DublinCores.java:186) ~[?:?]
	at org.opencastproject.metadata.dublincore.DublinCoreCatalogService.load(DublinCoreCatalogService.java:182) ~[?:?]
	at org.opencastproject.series.endpoint.SeriesRestService.addOrUpdateSeries(SeriesRestService.java:423) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179) ~[?:?]
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) ~[?:?]
	... 107 more

```
</details>

lkiesow commented:
To sum up our discussion about this:

- I won't fix the series service in here. That's a problem, but an unrelated one. I may address it later
- “Catalogs are expected to be XML”, the old Opencast developers said, “So this should fail!”. “But we already use non XML in production for some special cases”, the new dev answered. “Dammit!”, the old dev starts cursing. “I guess we can't break prod systems.”

I'll see to it that `addCatalog` is not being checked unless you ingest using a DC flavor (or something like that).

lkiesow commented:
@Arnei, I looked into this a bit more and it seems like `addCatalog` should only accept XML after all.
If you ingest a non-DC-catalog like this:

```bash
curl -f -u … "…/ingest/addCatalog" \
  -F "mediaPackage=<${TMP_MP}" \
  -F flavor=json-test/source \
  -F BODY=@test.json
```

Opencast will accept that right now, but it will also result in a media package (element) like this:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<mediapackage xmlns="http://mediapackage.opencastproject.org" id="bd5ef5b5-0757-47a7-9132-d322765758a7" start="2021-06-19T10:39:05Z">
  <media><!-- ... --></media>
  <metadata>
    <!-- ... -->
    <catalog id="ffc8d961-e5f7-47b5-bd83-aaf6e02b7a46" type="json-test/source">
      <mimetype>text/xml</mimetype>
      <tags/>
      <url>https://develop.opencast.org/files/mediapackage/bd5ef5b5-0757-47a7-9132-d322765758a7/ffc8d961-e5f7-47b5-bd83-aaf6e02b7a46/test.json</url>
    </catalog>
  </metadata>
  <attachments/>
  <publications/>
</mediapackage>
```

Note, that it says `<mimetype>text/xml</mimetype>`.

Let's discuss on Monday what we want to do about this…

lkiesow committed:
Temporarily Allow JSON Objects

This patch adds an exception for JSON arrays in Opencast 9, clearly
warning that this is not intended and will fail in future versions.

This will give people time to migrate while already catching most
issues.

lkiesow commented:
This patch now temporarily accepts JSON objects, but warns if they are ingested. Ad per discussion in the technical meeting, we want to prevent XML in the long term altogether, and I will create a followup which removes the exception again for later Opencast versions.

Arnei commented on modules/ingest-service-impl/src/main/java/org/opencastproject/ingest/impl/IngestServiceImpl.java:
```diff
@@ -918,15 +922,38 @@ public MediaPackage addCatalog(InputStream in, String fileName, MediaPackageElem
+      try (InputStream inputStream = workingFileRepository.get(mediaPackageId, elementId)) {
+        try (BufferedReader reader  = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
+          // Exception for current BBB integration which is ingesting a JSON object as catalog
+          isJSON = reader.read() == '{';
```
This fails for JsonArrays, which are expected by [Cut Marks To Smil](https://docs.opencast.org/r/10.x/admin/#workflowoperationhandlers/cut-marks-to-smil-woh.html).

lkiesow replied:
Ups… I was sure that was an object.
But then let's change this to

```java
isJSON = reader.read() == '[';
```

I'll do the update later today.

lkiesow replied:
Fixed
